File: src/Find_Minimum_in_Rotated_Sorted_Array.py
# -*- coding: utf-8 -*-


class Solution:
    # @param {integer[]} nums
    # @return {integer}
    def findMin(self, nums):
        l, r = 0, len(nums) - 1
        while l < r:
            mid = (l + r) >> 1
            if nums[l] <= nums[mid] <= nums[r]:
                return nums[l]
            elif nums[l] >= nums[mid] >= nums[r]:
                return nums[r]
            else:
                if nums[mid] < nums[l]:
                    r = mid
                else:
                    l = mid
        return nums[l]

if __name__ == "__main__":
    solution = Solution()
    print solution.findMin([4, 5, 6, 7, 0, 1, 2])
    print solution.findMin([1])

File: src/algorithms/cpp/Find_Minimum_in_Rotated_Sorted_Array.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <vector>
#include <map>
#include <set>
using namespace std;
#define out(v) cerr << #v << ": " << (v) << endl
#define SZ(v) ((int)(v).size())
const int maxint = -1u>>1;
template <class T> bool get_max(T& a, const T &b) {return b > a? a = b, 1: 0;}
template <class T> bool get_min(T& a, const T &b) {return b < a? a = b, 1: 0;}

class Solution {
public:
    int findMin(vector<int>& nums) {
        int l = 0, r = SZ(nums) - 1;
        int mid = -1;
        while (l < r - 1) {
            mid = (l + r) >> 1;
            if (nums[mid] > nums[r])    l = mid;
            else                        r = mid;
        }
        if (nums[l] < nums[r])  return nums[l];
        else                    return nums[r];
    }
};

int main() {
    vector <int> v1 = {4, 5, 6, 0, 1, 2};
    vector <int> v2 = {6, 0, 1, 2, 4, 5};
    vector <int> v3 = {0, 1, 2, 3, 4, 5};
    Solution solution = Solution();
    cout << solution.findMin(v1) << endl;
    cout << solution.findMin(v2) << endl;
    cout << solution.findMin(v3) << endl;
    return 0;
}
